zipline/data/data_portal.py

```diff
@@ -0,0 +1,475 @@
+#
+# Copyright 2015 Quantopian, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import bcolz
+from logbook import Logger
+
+import numpy as np
+import pandas as pd
+from pandas.tslib import normalize_date
+
+from zipline.assets import Future, Equity
+from zipline.data.us_equity_pricing import NoDataOnDate
+
+from zipline.utils import tradingcalendar
+from zipline.errors import (
+    NoTradeDataAvailableTooEarly,
+    NoTradeDataAvailableTooLate
+)
+
+log = Logger('DataPortal')
+
+BASE_FIELDS = {
+    'open': 'open',
+    'open_price': 'open',
+    'high': 'high',
+    'low': 'low',
+    'close': 'close',
+    'close_price': 'close',
+    'volume': 'volume',
+    'price': 'close'
+}
+
+
+class DataPortal(object):
+    def __init__(self,
+                 env,
+                 sim_params=None,
+                 equity_daily_reader=None,
+                 equity_minute_reader=None,
+                 future_daily_reader=None,
+                 future_minute_reader=None,
+                 adjustment_reader=None):
+        self.env = env
+
+        # Internal pointers to the current dt (can be minute) and current day.
+        # In daily mode, they point to the same thing. In minute mode, it's
+        # useful to have separate pointers to the current day and to the
+        # current minute.  These pointers are updated by the
+        # AlgorithmSimulator's transform loop.
+        self.current_dt = None
+        self.current_day = None
+
+        self.views = {}
+
+        self._asset_finder = env.asset_finder
+
+        self._carrays = {
+            'open': {},
+            'high': {},
+            'low': {},
+            'close': {},
+            'volume': {},
+            'sid': {},
+            'dt': {},
+        }
+
+        self._adjustment_reader = adjustment_reader
+
+        # caches of sid -> adjustment list
+        self._splits_dict = {}
+        self._mergers_dict = {}
+        self._dividends_dict = {}
+
+        # Cache of sid -> the first trading day of an asset, even if that day
+        # is before 1/2/2002.
+        self._asset_start_dates = {}
+        self._asset_end_dates = {}
+
+        # Handle extra sources, like Fetcher.
+        self._augmented_sources_map = {}
+        self._extra_source_df = None
+
+        self._sim_params = sim_params
```

Comment:
I don't think this needs to be an instance attribute.  It's never used for anything except getting the data frequency.